sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,577 @@
+        formatTokens.last != OptionalMinusSign())) {
+      notAtEndOfNumberFormatError(s"'$OPTIONAL_MINUS_STRING'")
+    }
+    // Make sure the format string contains at most one closing angle bracket at the end.
+    else if (inputTokenCounts.getOrElse(ClosingAngleBracket(), 0) > 1 ||
+      (inputTokenCounts.getOrElse(ClosingAngleBracket(), 0) == 1 &&
+        formatTokens.last != ClosingAngleBracket())) {
+      notAtEndOfNumberFormatError(s"'$WRAPPING_ANGLE_BRACKETS_TO_NEGATIVE_NUMBER'")
+    }
+    // Make sure that any dollar sign in the format string occurs before any digits.
+    else if (firstDigitIndex < firstDollarSignIndex) {
+      s"Currency characters must appear before digits in the number format: '$originNumberFormat'"
+    }
+    // Make sure that any dollar sign in the format string occurs before any decimal point.
+    else if (firstDecimalPointIndex != -1 &&
+      firstDecimalPointIndex < firstDollarSignIndex) {
+      "Currency characters must appear before any decimal point in the " +
+        s"number format: '$originNumberFormat'"
+    }
+    // Make sure that any thousands separators in the format string have digits before and after.
+    else if (digitGroupsBeforeDecimalPoint.exists {
+      case DigitGroups(tokens, _) =>
+        tokens.zipWithIndex.exists({
+          case (_: ThousandsSeparator, j: Int) if j == 0 || j == tokens.length - 1 =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j - 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j + 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case _ =>
+            false
+        })
+    }) {
+      "Thousands separators (,) must have digits in between them " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Thousands separators are not allowed after the decimal point, if any.
+    else if (digitGroupsAfterDecimalPoint.exists {
+      case DigitGroups(tokens, digits) =>
+        tokens.length > digits.length
+    }) {
+      "Thousands separators (,) may not appear after the decimal point " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Validation of the format string finished successfully.
+    else {
+      ""
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   *
+   * Iterates through the [[formatTokens]] obtained from processing the format string, while also
+   * keeping a parallel index into the input string. Throws an exception if the latter does not
+   * contain expected characters at any point.
+   *
+   * @param input the string that needs to converted
+   * @return the result Decimal value obtained from string parsing
+   */
+  def parse(input: UTF8String): Decimal = {
+    // Get a reference to the underlying input byte array without copying memory.
+    def inputBuffer: Array[Byte] = input.getBaseObject.asInstanceOf[Array[Byte]]
```

Comment:
This doesn't work with off-heap setting, and we also need to take care of Unicode where one char needs more than one byte. I think `input.toString` is better here.

Reply:
Done.